feat(routing-forms): replace FormEdit component with FormBuilder from event-types

oliverqx opened the pull request:
/claim #18987
## What does this PR do?

- Fixes #18987
- Fixes CAL-5097

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

visit routing-forms, edit or create new one.

oliverqx committed:
feat(routing-forms): replace FormEdit component with FormBuilder

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=19008) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=19008) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=19008) it.</sub>

oliverqx committed:
Merge branch 'main' into feat.routing-forms-edit-form-component

oliverqx committed:
fix: update FormBuilder.formProp bookingFields -> fields

parent form expects all form fields under the fields key, not bookingFields

oliverqx committed:
chore(form-builder): move BookingField type to form-builder schema file

oliverqx committed:
fix(form-builder): make disabled and LockedIcon props optional

Praashh commented on packages/app-store/routing-forms/pages/form-edit/[...appPages].tsx:
```diff
@@ -471,11 +476,37 @@ export default function FormEditPage({
-      Page={({ hookForm, form }) => <FormEdit appUrl={appUrl} hookForm={hookForm} form={form} />}
+      Page={({ hookForm, form }) => (
+        <>
+          {/* <FormEdit appUrl={appUrl} hookForm={hookForm} form={form} /> */}
```
```suggestion
```

Praashh reviewed (changes requested):
Can you please add test cases and a loom video for this feature?

oliverqx committed:
fix(routing-forms): remove adding a field in addForm test utility

should not have any side effects

oliverqx committed:
chore(form-builder): sync test fieldTypes with fieldTypesConfigMap

oliverqx commented:
just recorded the [loom video](https://www.loom.com/share/d813cd46dfe248de80f1c966bfe7d40c?sid=6e256881-aa4c-4bb3-a30e-ae72dad38179)

oliverqx committed:
fix: reinstate waitforSelector

oliverqx committed:
fix(form-builder): add className to SelectField

expected by some tests

oliverqx committed:
fix(routing-forms): modified verifySelectOptions

oliverqx committed:
fix(routing-forms): update test utils

1. remove filtering by hard coded values in favour of field.systemOnly
2. return options by value instead of label. return array is used to locate fields, and the test id generation uses values, not label. Included comments to explain my reasoning

oliverqx committed:
fix(routing-forms): sync addAllTypesOfFieldsAndSaveForm to FormBuilder's dynamic

function used to test the previous dynamic, where an accordion was rendered to create and edit fields. Now its a dialog, and data-testids have changed.

oliverqx committed:
fix(routing-forms): upon form submission add id if not already present

FormBuilder does not generate ids for fields, but Routing Forms expects an id.

oliverqx committed:
Merge branch 'main' into feat.routing-forms-edit-form-component

oliverqx committed:
feat(routing-forms): add support for url field

oliverqx committed:
feat(routing-forms): add support for address field

oliverqx commented:
do you guys think its possible to increase the bounty amount? 

Just realized that to have only one set of custom components as requested, there are multiple pieces of the code that I need to understand deeply in order to make the changes work without any side effects/issues. 

Namely, the Form Builder and then Routing Forms which includes react-awesome-query-builder and its flow. 

Then also updating each of the tests in Routing Forms because test selectors are widely different between current Routing Forms implementation and the Form Builder.

Already invested around 7 hours into it and would appreciate an adjustment to the bounty

cc: @PeerRich @Praashh @hariombalhara

oliverqx committed:
fix: remove dead code

oliverqx committed:
fix: updaate zod schema for options

id is not set by FormBuilder

oliverqx committed:
feat(routing-forms): create URL widget

also updates FormBuilder Components to use URL Widget

oliverqx committed:
feat(routing-forms): add multi email support

oliverqx committed:
feat(routing-forms): add support for radio group

oliverqx committed:
feat(routing-forms): add checkbox group support

oliverqx committed:
feat(routing-forms): add boolean support

oliverqx committed:
chore(form-builder): update Components to use Widgets

oliverqx committed:
feat: update phone widget

oliverqx committed:
fix(routing-forms): fix test suite WIP

oliverqx committed:
fix: tests

oliverqx committed:
Merge branch 'main' into feat.routing-forms-edit-form-component

oliverqx committed:
fix: update type

oliverqx committed:
feat: fix type

oliverqx committed:
fix: update types

oliverqx committed:
fix: update type

oliverqx committed:
fix: update type

oliverqx commented:
@hariombalhara @Praashh the pr is ready to be reviewed. I fixed test suite for routing forms, now that form builder is being used dynamic is a bit different when it comes to identifiers and test-ids

oliverqx committed:
chore: remove old form field editor

oliverqx committed:
temp: test crypto mock

oliverqx committed:
temp: test

oliverqx committed:
fix: tests

oliverqx committed:
fix: update test

oliverqx committed:
Merge branch 'main' into feat.routing-forms-edit-form-component

oliverqx committed:
Merge branch 'main' into feat.routing-forms-edit-form-component

oliverqx commented:
@hariombalhara @Praashh #19473, insane

oliverqx commented:
cc: @CarinaWolli @PeerRich

PeerRich commented:
/bonus 100

PeerRich commented:
@anikdhabal can you take over the review since Praashh is no with us longer reviewing? 🙏

hariombalhara commented on packages/app-store/routing-forms/playwright/tests/basic.e2e.ts:
```diff
@@ -178,84 +167,78 @@ test.describe("Routing Forms", () => {
-      expect(url.searchParams.get("Test Field Phone")).toBe("456");
-      expect(url.searchParams.get("Test Field Email")).toBe("456@example.com");
-    });
+    // test("should be able to submit a prefilled form with all types of fields", async ({ page }) => {
```
We need to ensure that these tests pass

oliverqx replied:
oh sorry i commented them out when running the other tests, will uncomment.

hariombalhara commented on packages/app-store/routing-forms/pages/form-edit/[...appPages].tsx:
```diff
@@ -50,431 +38,40 @@ const appendArray = <T,>({
-      Page={({ hookForm, form }) => <FormEdit appUrl={appUrl} hookForm={hookForm} form={form} />}
+      Page={({ hookForm, form }) => (
+        <>
+          <FormBuilder
```
This is great to see that it reduced the file size a lot !!

But we need to ensure that all functionalities that were supported in Form-edit work. If there is some not-critical functionality that doesn't work, let's make a list and then we can decide if it can be fixed in a followup or required to be fixed in this PR itself

hariombalhara commented on packages/app-store/routing-forms/pages/form-edit/[...appPages].tsx:
```diff
@@ -50,431 +38,40 @@ const appendArray = <T,>({
       appUrl={appUrl}
-      Page={({ hookForm, form }) => <FormEdit appUrl={appUrl} hookForm={hookForm} form={form} />}
+      Page={({ hookForm, form }) => (
+        <>
```
I don't think we need this Fragment.

hariombalhara commented on packages/app-store/routing-forms/pages/form-edit/[...appPages].tsx:
```diff
@@ -50,431 +38,40 @@ const appendArray = <T,>({
+            }}
+            shouldConsiderRequired={(field: BookingField) => {
+              // Location field has a default value at backend so API can send no location but we don't allow it in UI and thus we want to show it as required to user
+              return field.name === "location" ? true : field.required;
```
There is no location field in Routing Form. So, let's remove this for now from here.

hariombalhara commented on packages/app-store/routing-forms/pages/form-edit/[...appPages].tsx:
```diff
@@ -50,431 +38,40 @@ const appendArray = <T,>({
+              options: {
+                locations: {
+                  // FormBuilder doesn't handle plural for non-english languages. So, use english(Location) only. This is similar to 'Workflow'
+                  source: { label: "Location" },
```
Same here.

hariombalhara commented:
Things that must be tested
- Prefill of Routing Form fields
- Submission of Routing Form
- Responses must be auto-forwarded as query params if 'Send Booker to' is set as some event-type

hariombalhara commented:
@oliverqx Could you please record a loom or similar to show how it functions now?

retrogtx commented:
@oliverqx is this still on?

if not, will close since this is stale

oliverqx commented:
@retrogtx will be finishing work on it this weekend

retrogtx commented:
hey @oliverqx
a gentle ping

anikdhabal commented:
@oliverqx could you pls resolve the conflicts and finish the work

oliverqx commented:
will not address this pr any longer as it consumed too much time already given the bounty.
